File: app/models/concerns/error_trackable.rb
# ErrorTrackable Concern
# Adds error tracking and analytics capabilities to models

module ErrorTrackable
  extend ActiveSupport::Concern
  
  included do
    # Track error patterns
    scope :with_errors, -> { where.not(error_code: nil) }
    scope :without_errors, -> { where(error_code: nil) }
  end
  
  class_methods do
    # Get error statistics
    def error_stats
      with_errors.group(:error_code).count.sort_by { |_, count| -count }
    end
    
    # Get most common errors (top N)
    def top_errors(limit = 10)
      error_stats.take(limit)
    end
    
    # Get error rate
    def error_rate
      total = count
      return 0 if total == 0
      
      errors = with_errors.count
      (errors.to_f / total * 100).round(2)
    end
    
    # Group errors by category
    def errors_by_category
      categories = {
        invalid_format: [],
        not_found: [],
        authentication: [],
        rate_limit: [],
        network: [],
        other: []
      }
      
      with_errors.pluck(:error_code).each do |error|
        next if error.blank?
        
        case error.downcase
        when /invalid|format|malformed/
          categories[:invalid_format] << error
        when /not found|does not exist/
          categories[:not_found] << error
        when /auth|permission|unauthorized/
          categories[:authentication] << error
        when /rate limit|too many/
          categories[:rate_limit] << error
        when /network|timeout|connection/
          categories[:network] << error
        else
          categories[:other] << error
        end
      end
      
      # Return with counts
      categories.transform_values(&:size).select { |_, count| count > 0 }
    end
  end
  
  # Instance methods
  def has_error?
    error_code.present?
  end
  
  def error_category
    return nil unless has_error?
    
    case error_code.downcase
    when /invalid|format|malformed/
      :invalid_format
    when /not found|does not exist/
      :not_found
    when /auth|permission|unauthorized/
      :authentication
    when /rate limit|too many/
      :rate_limit
    when /network|timeout|connection/
      :network
    else
      :other
    end
  end
  
  def error_severity
    return :none unless has_error?
    
    case error_category
    when :authentication, :rate_limit
      :critical
    when :invalid_format, :not_found
      :low
    when :network
      :medium
    else
      :medium
    end
  end
  
  def error_recoverable?
    return false unless has_error?
    
    case error_category
    when :network, :rate_limit
      true
    when :invalid_format, :not_found, :authentication
      false
    else
      false  # Conservative: don't retry unknown errors
    end
  end
end

File: laravel-app/app/Traits/ErrorTrackable.php
<?php

namespace App\Traits;

use Illuminate\Database\Eloquent\Builder;

/**
 * ErrorTrackable Trait
 * Adds error tracking and analytics capabilities to models
 */
trait ErrorTrackable
{
    /**
     * Query Scopes
     */
    public function scopeWithErrors(Builder $query): Builder
    {
        return $query->whereNotNull('error_code');
    }

    public function scopeWithoutErrors(Builder $query): Builder
    {
        return $query->whereNull('error_code');
    }

    /**
     * Class Methods - Error Statistics
     */
    public static function errorStats(): array
    {
        return self::withErrors()
                   ->groupBy('error_code')
                   ->selectRaw('error_code, COUNT(*) as count')
                   ->orderByDesc('count')
                   ->pluck('count', 'error_code')
                   ->toArray();
    }

    public static function topErrors(int $limit = 10): array
    {
        $stats = self::errorStats();
        return array_slice($stats, 0, $limit, true);
    }

    public static function errorRate(): float
    {
        $total = self::count();
        
        if ($total === 0) {
            return 0;
        }

        $errorsCount = self::withErrors()->count();
        
        return round(($errorsCount / $total) * 100, 2);
    }

    public static function errorsByCategory(): array
    {
        $categories = [
            'invalid_format' => 0,
            'not_found' => 0,
            'authentication' => 0,
            'rate_limit' => 0,
            'network' => 0,
            'other' => 0,
        ];

        $errors = self::withErrors()->pluck('error_code');

        foreach ($errors as $error) {
            if (empty($error)) {
                continue;
            }

            $errorLower = strtolower($error);
            
            if (preg_match('/(invalid|format|malformed)/', $errorLower)) {
                $categories['invalid_format']++;
            } elseif (preg_match('/(not found|does not exist)/', $errorLower)) {
                $categories['not_found']++;
            } elseif (preg_match('/(auth|permission|unauthorized)/', $errorLower)) {
                $categories['authentication']++;
            } elseif (preg_match('/(rate limit|too many)/', $errorLower)) {
                $categories['rate_limit']++;
            } elseif (preg_match('/(network|timeout|connection)/', $errorLower)) {
                $categories['network']++;
            } else {
                $categories['other']++;
            }
        }

        // Return only categories with errors
        return array_filter($categories, fn($count) => $count > 0);
    }

    /**
     * Instance Methods
     */
    public function hasError(): bool
    {
        return !empty($this->error_code);
    }

    public function errorCategory(): ?string
    {
        if (!$this->hasError()) {
            return null;
        }

        $errorLower = strtolower($this->error_code);

        if (preg_match('/(invalid|format|malformed)/', $errorLower)) {
            return 'invalid_format';
        } elseif (preg_match('/(not found|does not exist)/', $errorLower)) {
            return 'not_found';
        } elseif (preg_match('/(auth|permission|unauthorized)/', $errorLower)) {
            return 'authentication';
        } elseif (preg_match('/(rate limit|too many)/', $errorLower)) {
            return 'rate_limit';
        } elseif (preg_match('/(network|timeout|connection)/', $errorLower)) {
            return 'network';
        } else {
            return 'other';
        }
    }

    public function errorSeverity(): string
    {
        if (!$this->hasError()) {
            return 'none';
        }

        return match ($this->errorCategory()) {
            'authentication', 'rate_limit' => 'critical',
            'invalid_format', 'not_found' => 'low',
            'network' => 'medium',
            default => 'medium',
        };
    }

    public function errorRecoverable(): bool
    {
        if (!$this->hasError()) {
            return false;
        }

        $category = $this->errorCategory();

        // Network and rate limit errors are typically recoverable
        if (in_array($category, ['network', 'rate_limit'])) {
            return true;
        }

        // Invalid format, not found, and authentication errors are not recoverable
        if (in_array($category, ['invalid_format', 'not_found', 'authentication'])) {
            return false;
        }

        // Conservative: don't retry unknown errors
        return false;
    }

    /**
     * Get a human-readable error message
     */
    public function errorMessageFormatted(): ?string
    {
        if (!$this->hasError()) {
            return null;
        }

        $category = $this->errorCategory();
        $severity = $this->errorSeverity();

        return match ($category) {
            'invalid_format' => "Invalid format: {$this->error_code}",
            'not_found' => "Not found: {$this->error_code}",
            'authentication' => "Authentication error: {$this->error_code}",
            'rate_limit' => "Rate limit exceeded: {$this->error_code}",
            'network' => "Network error: {$this->error_code}",
            default => "Error: {$this->error_code}",
        };
    }

    /**
     * Check if error should be retried
     */
    public function shouldRetryError(): bool
    {
        return $this->hasError() && $this->errorRecoverable();
    }
}
